File: backend/app/policies/user_policy.rb
class UserPolicy < ApplicationPolicy
  def index?
    user.admin? || user.super_admin?
  end

  def show?
    user.admin? || user.super_admin? || user == record
  end

  def create?
    user.super_admin?
  end

  def update?
    user.super_admin? || user == record
  end

  class Scope < Scope
    def resolve
      if user.admin? || user.super_admin?
        scope.all
      else
        scope.where(id: user.id)
      end
    end
  end
end

File: backend/app/Policies/UserPolicy.ts
import { BasePolicy } from '@ioc:Adonis/Addons/Bouncer'
import User from 'App/Models/User'

export default class UserPolicy extends BasePolicy {
  public async viewList(user: User) {
    const roles = await user.related('roles').query()
    return roles.some((role) => role.slug === 'admin')
  }

  public async view(user: User) {
    const roles = await user.related('roles').query()
    return roles.some((role) => role.slug === 'admin')
  }

  public async create(user: User) {
    const roles = await user.related('roles').query()
    return roles.some((role) => role.slug === 'admin')
  }

  public async update(user: User) {
    const roles = await user.related('roles').query()
    return roles.some((role) => role.slug === 'admin')
  }

  public async delete(user: User) {
    const roles = await user.related('roles').query()
    return roles.some((role) => role.slug === 'admin')
  }
}
